docs: Adds RBAC intro & guides (#19222)
MaterializeInc/materialize: doc/user/content/manage/user-management/rbac-newrole.md

---
title: "Create a new user and role"
description: ""
menu:
  main:
    parent: user-management
    weight: 15
---

This guide walks you through creating a new user and new role in Materialize. By
the end of this tutorial you will:

* Invite a new user in the Materialize Console
* Create a new role in your Materialize account using `psql`
* Apply privileges to the new role
* Assign the new role to the new user

## Before you begin

* Make sure you have `psql` installed locally.

* Make sure you have a [Materialize account](https://materialize.com/register/?utm_campaign=General&utm_source=documentation).

## Step 1. Invite a new user

> First, [login to the Materialize console](https://console.materialize.com/) and navigate to Account > Account

Reviewer: ```suggestion
1. [Login to the Materialize console](https://console.materialize.com/) and navigate to Account > Account
```

Settings > Users.

> Click **Invite User** and fill in the user information.

Reviewer: ```suggestion
1. Click **Invite User** and fill in the user information.
```

Click **Invite User** and fill in the user information.

The **Organization Admin** and **Organization Member** roles refer to console

> privileges and are not related to RBAC roles.

Reviewer 2: **Organization Admin** will cause the role in Materialize to be created with superuser privileges.

Materialize stores role names, IDs, and object IDs in the `mz_roles` catalog
table. You need to connect to your Materialize instance with `psql` to add roles
to this table.

> 1. In the [Materialize UI](https://console.materialize.com/), go to the **Connect** screen, create a new app password and then copy the `psql` command.

Reviewer: ```suggestion
1. In the [Materialize UI](https://console.materialize.com/), go to the **Connect** screen.

1. Create a new app password.

    The app password will be displayed only once, so be sure to copy the password somewhere safe. If you forget your password, you can create a new one.

1. Copy the `psql` command.
```

1. In the [Materialize UI](https://console.materialize.com/), enable the region where you want to run Materialize.

1. On the **Connect** screen, create a new app password and then copy the `psql` command.

    The app password will be displayed only once, so be sure to copy the password somewhere safe. If you forget your password, you can create a new one.

1. Open a new terminal window, run the `psql` command, and enter your app password.

    Now that you're logged in, you can create a new role in the `mz_roles`

>     table.

Reviewer 2: It's probably worth mentioning that by logging in for the first time, a new role is automatically created with the same name as the frontegg account name.

    CREATE ROLE dev_role;
    ```

1. Each role you create has default **role attributes** that determine how they
    can interact with Materialize objects. Let's look at the role attributes of
    the role you created:

    ```sql
    SELECT * FROM mz_roles WHERE 'name' = dev_role;
    ```

    Your `dev_role` returns attributes similar to the following:

    ```nofmt
    -[ RECORD 1 ]--+------
    id             | u8
    oid            | 50991
    name           | dev_role
    inherit        | t
    create_role    | f
    create_db      | f
    create_cluster | f
    ```
    Your `id` and `oid` values will look different.

    The `inherit`, `create_role`, `create_db`, and `create_cluster` are the
    role attributes assigned to a role when it is created. These attributes
    determine the system-level permissions of a role and do not impact object-level privileges.

>    * `INHERIT` is set to true by default and allows roles to inherit the

Reviewer: This isn't rendering right:
<img width="764" alt="Screenshot 2023-05-18 at 10 16 26 AM" src="https://github.com/MaterializeInc/materialize/assets/15893490/d5b25168-21bf-4e77-8baa-1a25b1d4c268">

    `INHERIT` is set to true by default and allows roles to inherit the
    privileges of roles it is a member of.

    `CREATEROLE` allows the role to create, change, or delete other roles or
    assign role membership.

    `CREATEDB` allows the role to create new databases.

    `CREATECLUSTER` allows the role to create Materialize clusters. This
    attribute is unique to the Materialize concept of clusters.

## Create example objects

Your `dev` role has the default system level permissions and needs object
specific privileges. RBAC allows you to apply granular privileges to objects in the SQL hierarchy. Let's create some example objects in the system and determine what
privileges the role needs.

1. In your `psql` terminal, create a new example cluster to avoid impacting
   other environments.

   ```sql
   CREATE CLUSTER devcluster;
   ```

1. Change into the example cluster.

   ```sql
   SET CLUSTER TO devcluster;
   ```

> 1. Create a new database, schema, and table in the cluster.

Reviewer 2: These objects aren't being created in a cluster.

   ```sql
   CREATE DATABASE devdb;
   ```

   ```sql
   CREATE SCHEMA devdb.schema;
   ```

   ```sql
   CREATE TABLE d (a int, b text NOT NULL);
   ```

You just created a nested hierarchy of objects. Your schema object belongs to
the database and the database is only accessible within the cluster. The next
step is to grant privileges to your role based on the role needs.

## Grant privileges to a role

In this example, let's say your `dev` role needs the following permissions:

* Read, write, and append privileges on the table
* Update privileges on the schema
* All available privileges on the database

1. In your terminal, grant table-level privileges to the `dev` role.

   ```sql
   GRANT SELECT, UPDATE, INSERT ON d TO dev_role;
   ```

   Table objects have four available privileges - `read`, `write`, `insert`, and
   `delete`. The `dev` role doesn't need `delete` permissions, so it is not
   applied in the GRANT statement above.

2. Grant schema privileges to the `dev` role.

   ```sql
   GRANT USAGE, CREATE ON SCHEMA devdb.schema TO dev_role;
   ```

>    Schemas have `UPDATE` and `CREATE` privileges available to grant.

Reviewer 2: `USAGE` not `UPDATE`.

3. Grant database privileges to the `dev` role.

   ```sql
   GRANT USAGE, CREATE ON DATABASE devdb TO dev_role;
   ```

4. Grant cluster privileges to the `dev` role.

   ```sql
   GRANT USAGE, CREATE ON CLUSTER devcluster TO dev_role;
   ```

   Materialize cluster privileges are unique to the Materialize RBAC structure.
   To have access to the objects within a cluster, you must also have the same
   level of access to the cluster itself.

## Assign a role to a user

The `dev` role now has the acceptable privileges it needs. Let's apply this role
to a user in your Materialize organization.

1. In your terminal, use the `GRANT` statement to apply a role to your new user.

   ```sql
   GRANT dev TO <new_user>;
   ```

1. To review the permissions a role has, you can view the object data.

   ```sql

>    SELECT name, privileges::text FROM mz_tables WHERE name='t';

Reviewer 2: I'm not sure if `privileges::text` is right here, what does an example output look like?

Reviewer 2: Oh I see the output below. I didn't realize that this worked, that's neat.

Author: Ah that was actually meant to be a placeholder and I did not mark it well. This actually gives me `ERROR:  CAST does not support casting from mz_aclitem[] to text[]`

>    SELECT name, privileges::text FROM mz_tables WHERE name='t';

Reviewer 2: Above the table is named `d`.

   ```

   The output should return the owner, the level of permission, and the name of
   the role on the table.

   ```shell
   name|privileges
   t|{admin@mz.com=arwd/admin@mz.com,dev_role=arw/admin@mz.com}
   (1 row)
   ```

   In this example, the administrator has append, read, write, and delete
   privileges on the table. The `dev` role has append, read, and write privileges,
   which were assigned by the admin user.

## Next Steps

You just used RBAC to manage user privileges in Materialize! The next guide will
cover altering and dropping user roles. Move to the next guide in the series, or
destroy the objects you created for this guide.

```
DROP CLUSTER devcluster CASCADE;

> ```

Reviewer 2: You should add `DROP DATABASE devdb CASCADE` too.

Reviewer 2: Also `DROP TABLE d`

   `delete`. The `dev_role` doesn't need `delete` permissions, so it is not
   applied in the `GRANT` statement above.

2. Grant schema privileges to the `dev_role`:

   ```sql
   GRANT USAGE, CREATE ON SCHEMA dev_db.schema TO dev_role;
   ```

   Schemas have `USAGE` and `CREATE` privileges available to grant.

3. Grant database privileges to the `dev_role`:

   ```sql
   GRANT USAGE, CREATE ON DATABASE dev_db TO dev_role;
   ```

4. Grant cluster privileges to the `dev_role`:

   ```sql
   GRANT USAGE, CREATE ON CLUSTER dev_cluster TO dev_role;
   ```

   Materialize cluster privileges are unique to the Materialize RBAC structure.
   To have access to the objects within a cluster, you must also have the same
   level of access to the cluster itself.

## Step 6. Assign a role to a user

The `dev_role` now has the acceptable privileges it needs. Let's apply this role
to a user in your Materialize organization.

1. In your terminal, use the `GRANT` statement to apply a role to your new user:

   ```sql
   GRANT dev_role TO <new_user>;
   ```

1. To review the permissions a role has, you can view the object data:

   ```sql

>    SELECT name, privileges::text FROM mz_tables WHERE name='dev_table';

Author: ```suggestion
   SELECT name, privileges FROM mz_tables WHERE name='dev_table';
```

   ```

   The output should return the owner, the level of permission, and the name of
   the role on the table.

   ```nofmt

>    name|privileges
>    t|{admin@mz.com=arwd/admin@mz.com,dev_role=arw/admin@mz.com}
>    (1 row)

Author: ```suggestion
   name|privileges
   dev_table|{u1=UC/u1,u8=UC/u1}
   (1 row)
```

## Before you begin

In this guide, you'll use `psql` to manage users in Materialize, so be sure you
have `psql` installed locally.

If you have not signed up for a free trial, [sign up now](https://materialize.com/register/?utm_campaign=General&utm_source=documentation).

## Invite a new user

First, login to your Materialize console and navigate to Account > Account
Settings > Users.

Click **Invite User** and fill in the user information.

The **Organization Admin** and **Organization Member** roles refer to `SUPERUSER`
privileges

## Create a new role

Materialize stores role names, IDs, and object IDs in the `mz_roles` catalog
table. You need to connect to your Materialize instance with `psql` to add roles
to this table.

1. In the [Materialize UI](https://console.materialize.com/), enable the region where you want to run Materialize.

. On the **Connect** screen, create a new app password and then copy the `psql` command.

    The app password will be displayed only once, so be sure to copy the password somewhere safe. If you forget your password, you can create a new one.

1. Open a new terminal window, run the `psql` command, and enter your app password.

>     Now that you're logged in, you can create a new role in the `mz_roles`

Reviewer: Looks like this is resolved but with no change...

>     Now that you're logged in, you can create a new role in the `mz_roles`

Author: Hm, thought I added these edits. Going back and checking again.

1. To review the permissions a role has, you can view the object data:

   ```sql
   SELECT name, privileges FROM mz_tables WHERE name='dev_table';
   ```

>    The output should return the owner object ID, the level of permission, and the owner ID of the role on the table.

Reviewer 2: You've specified owner ID twice here, did you only mean to include it once? Also you will need to add the `owner_id` column to your query to see the owner of a table.

   ```nofmt
   name|privileges
   dev_table|{u1=arwd/u1,u8=arw/u1}
   (1 row)
   ```

   In this example, object ID `u1` has append, read, write, and delete

>    privileges on the table. Object ID `u8` is the `dev_role` and has append, read, and write privileges,

Reviewer 2: "object ID" should be replaced with "role ID".
